File: C/aes.c
// * ECB of a 128-bit block
// ** Prior to first round
// *** generate round keys with key schedule
// *** add zeroth round key to block
// ** Main rounds
// *** SubBytes
// *** ShiftRows
// *** MixColumns
// *** AddRoundKey
// ** Final round
// *** SubBytes
// *** ShiftRows
// *** AddRoundKey

#include "aes.h"

u8 ltable[256], i_ltable[256], Sbox[256], i_Sbox[256];

void init_tables();
void init_Sbox();

/* --- Galois -- */

u8 gmul(u8 a, u8 b)
{
    if (!(a && b)) return 0;

    init_tables();
    return i_ltable[(ltable[a] + ltable[b]) % 0xff];
}

u8 ginv(u8 a)
{
    if (!a) printf("called ginv(0)\n"), exit(-1);

    init_tables();
    return i_ltable[0xff - ltable[a]];
}

u8 gdiv(u8 a, u8 b)
{
    init_tables();
    return gmul(a, ginv(b));
}

void init_tables()
{
    static u8 initialized;
    if (initialized) return;
    else initialized = 1;

    u8 a = 1, p = 0;

    // 3 is used as the generator
    for (u16 i = 0; i <= 0xff; i++) {
        i_ltable[i] = a;
        ltable[a] = i;

        a ^= (a << 1) ^ ((a & 0x80) ? 0x1b : 0);
    }

    i_ltable[0xff] = i_ltable[0];
    ltable[0] = 0;

    init_Sbox();

}

void init_Sbox()
{
    for (u16 i = 1; i <= 0xff; i++) {}
}


u32 rot_word(u32 word)
{
    return (word << 24) | (word >> 8);
}

u32 sub_word(u32 word)
{
    u8 b1 = (0xff000000 & word) >> 24;
    u8 b2 = (0x00ff0000 & word) >> 16;
    u8 b3 = (0x0000ff00 & word) >> 8;
    u8 b4 = 0x000000ff & word;

    return (Sbox[b1] << 24) | (Sbox[b2] << 16) | (Sbox[b3] << 8) | Sbox[b4];
}

u32* expand_key(u32* key)
{
    size_t num_rounds = 11;
    u32* rkeys = malloc(num_rounds * 16);

    u32 rcons[10];
    rcons[0] = 1;

    for (int i = 1; i < 10; i++)
        rcons[i] = gmul(rcons[i - 1], 2);

    for (int i = 0; i < 4 * num_rounds; i++) {
        if (i < 4)
            rkeys[i] = key[i];
        else if (i % 4 == 0)
            rkeys[i] = rkeys[i - 4] ^ sub_word(rot_word(rkeys[i - 1])) ^ rcons[i/4 - 1];
        else
            rkeys[i] = rkeys[i - 4] ^ rkeys[i - 1];
    }

    return rkeys;
}

/* --- Core --- */

void shift_rows(u8* block)
{
    for (int i = 0; i < 4; i++)
        for (int j = 0; j < i; j++) {
            u8 temp = block[i];
            block[i] = block[4 + i];
            block[4 + i] = block[8 + i];
            block[8 + i] = block[12 + i];
            block[12 + i] = temp;
        }
}

void i_shift_rows(u8* block)
{
    for (int i = 0; i < 4; i++)
        for (int j = 0; j < i; j++) {
            u8 temp = block[12 + i];
            block[12 + i] = block[8 + i];
            block[8 + i] = block[4 + i];
            block[4 + i] = block[i];
            block[i] = temp;
        }
}

void mix_columns(u8* block)
{
    for (u8 i = 0; i < 4; i++) {
        u8* col = block + 4*i;
        u8 a[4], b[4];

        for (u8 j = 0; j < 4; j++) {
            a[j] = col[j];
            b[j] = gmul(col[j], 2);
        }

        col[0] = b[0] ^ a[3] ^ a[2] ^ b[1] ^ a[1];
        col[1] = b[1] ^ a[0] ^ a[3] ^ b[2] ^ a[2];
        col[2] = b[2] ^ a[1] ^ a[0] ^ b[3] ^ a[3];
        col[3] = b[3] ^ a[2] ^ a[1] ^ b[0] ^ a[0];
    }
}

void i_mix_columns(u8* block)
{
    for (u8 i = 0; i < 4; i++) {
        u8* col = block + 4*i;
        u8 a[4];

        for (u8 i = 0; i < 4; i++)
            a[i] = col[i];

        col[0] = gmul(a[0],14) ^ gmul(a[2],13) ^ gmul(a[1],11) ^ gmul(a[3],9);
        col[1] = gmul(a[1],14) ^ gmul(a[3],13) ^ gmul(a[2],11) ^ gmul(a[0],9);
        col[2] = gmul(a[2],14) ^ gmul(a[0],13) ^ gmul(a[3],11) ^ gmul(a[1],9);
        col[3] = gmul(a[3],14) ^ gmul(a[0],11) ^ gmul(a[1],13) ^ gmul(a[2],9);
    }
}

void enc_block(u8* block, u8* key)
{
    u8* rkeys = (u8*)expand_key((u32*)key);

    // initial AddRoundKey
    for (u8 i = 0; i < 16; i++)
        block[i] ^= rkeys[i];

    // 10 core rounds (final round without MixColumns)
    for (u8 round = 1; round < 11; round++) {
        // SubBytes
        for (u8 i = 0; i < 16; i++)
            block[i] = Sbox[block[i]];

        // ShiftRows
        shift_rows(block);

        // MixColumns
        if (round != 10) mix_columns(block);

        // AddRoundKey
        for (u8 i = 0; i < 16; i++)
            block[i] ^= rkeys[16*round + i];
    }
}

void dec_block(u8* block, u8* key)
{
    u8* rkeys = (u8*)expand_key((u32*)key);

    // 10 core rounds (final round without MixColumns)
    for (u8 round = 10; round > 0; round--) {
        // AddRoundKey
        for (u8 i = 0; i < 16; i++)
            block[i] ^= rkeys[16*round + i];

        // MixColumns
        if (round != 10) i_mix_columns(block);

        // ShiftRows
        i_shift_rows(block);

        // SubBytes
        for (u8 i = 0; i < 16; i++)
            block[i] = i_Sbox[block[i]];
    }

    // initial AddRoundKey
    for (u8 i = 0; i < 16; i++)
        block[i] ^= rkeys[i];
}

u8 pad_to_block(u8* plain, u16 num_bytes)
{
    u8 num_blocks = num_bytes / 16 + ((num_bytes % 16) ? 1 : 0);
    u8 num_padding = 16*num_blocks - num_bytes;

    plain = realloc(plain, 16*num_blocks);

    for (u16 i = num_bytes; i < 16*num_blocks; i++)
        plain[i] = num_padding;

    return num_blocks;
}

/* --- Cipher Modes --- */

void ecb_enc(u8* plaintext, u8 num_blocks, u8* key)
{
    for (u8 i = 0; i < num_blocks; i++)
        enc_block(plaintext + 16*i, key);
}

void ecb_dec(u8* ciphertext, u8 num_blocks, u8* key)
{
    for (u8 i = 0; i < num_blocks; i++)
        dec_block(ciphertext + 16*i, key);
}

void cbc_enc(u8* plaintext, u8 num_blocks, u8* key, u8* iv)
{
    for (u8 block = 0; block < num_blocks; block++) {
        if (!block)
            for (u8 i = 0; i < 16; i++)
                plaintext[i] ^= iv[i];
        else
            for (u8 i = 0; i < 16; i++)
                plaintext[16*block + i] ^= plaintext[16*(block - 1) + i];

        enc_block(plaintext + 16*block, key);
    }
}

void cbc_dec(u8* ciphertext, u8 num_blocks, u8* key, u8* iv)
{
    for (u8 block = num_blocks - 1; block > 0; block--) {
        dec_block(ciphertext + 16*block, key);

        for (u8 i = 0; i < 16; i++)
            ciphertext[16*block + i] ^= ciphertext[16*(block - 1) + i];
    }

    dec_block(ciphertext, key);

    for (u8 i = 0; i < 16; i++)
        ciphertext[i] ^= iv[i];
}

File: python/aes.py
#!/usr/bin/env python3

log_lut =   [0] * 256
exp_lut =   [0] * 256
Sbox =      [0] * 256
inv_Sbox =  [0] * 256

tables_initialized = False

def init_tables(): # using 3 as a generator for GF(2^8)
    global tables_initialized
    if tables_initialized: return
    else: tables_initialized = True

    a = 1
    for i in range(256):
        exp_lut[i] = a # 3^i = a
        log_lut[a] = i # log3(a) = i

        a ^= (a << 1) ^ (0x1b if (a & 0x80) else 0) # a *= 3
        a &= 0xff

    init_Sbox()

def init_Sbox():
    def rotl_byte(a: int, n: int) -> int:
        for _ in range(n):
            hi_bit = a & 0x80
            a = (a << 1) & 0xff
            a |= hi_bit >> 7

        return a

    for n in range(1, 256):
        inv = ginv(n)
        Sbox[n] =   inv ^ \
                    rotl_byte(inv, 1) ^ \
                    rotl_byte(inv, 2) ^ \
                    rotl_byte(inv, 3) ^ \
                    rotl_byte(inv, 4) ^ \
                    0x63

        inv_Sbox[Sbox[n]] = n

    Sbox[0] = 0x63
    inv_Sbox[0x63] = 0

### Field Ops ###

def gmul(a: int, b: int) -> int:
    init_tables()

    if not a or not b: return 0
    return exp_lut[(log_lut[a] + log_lut[b]) % 0xff]

def ginv(a: int) -> int:
    if not a:
        raise ValueError('0 has no multiplicative inverse')

    init_tables()
    return exp_lut[0xff - log_lut[a]]

def gdiv(a: int, b: int) -> int:
    if not b:
        raise ValueError('division by 0')

    init_tables()
    return exp_lut[(log_lut[a] - log_lut[b]) % 0xff]

### Transformations ###

def subBytes(block: bytearray):
    init_tables()

    for i, b in enumerate(block):
        block[i] = Sbox[b]

def isubBytes(block: bytearray):
    init_tables()

    for i, b in enumerate(block):
        block[i] = inv_Sbox[b]

def shiftRows(block: bytearray):
    for r in range(4):
        row = [block[4*c + r] for c in range(4)]
        row = row[r:] + row[:r]

        for c in range(4):
            block[4*c + r] = row[c]

def ishiftRows(block: bytearray):
    for r in range(4):
        row = [block[4*c + r] for c in range(4)]
        row = row[-r:] + row[:-r]

        for c in range(4):
            block[4*c + r] = row[c]

def mixColumns(block: bytearray):
    for i in range(4):
        col = block[4*i:4*i + 4]
        xformed = 4 * [0]

        xformed[0] = gmul(col[0], 2) ^ gmul(col[1], 3) ^ gmul(col[2], 1) ^ gmul(col[3], 1)
        xformed[1] = gmul(col[0], 1) ^ gmul(col[1], 2) ^ gmul(col[2], 3) ^ gmul(col[3], 1)
        xformed[2] = gmul(col[0], 1) ^ gmul(col[1], 1) ^ gmul(col[2], 2) ^ gmul(col[3], 3)
        xformed[3] = gmul(col[0], 3) ^ gmul(col[1], 1) ^ gmul(col[2], 1) ^ gmul(col[3], 2)

        block[4*i:4*i + 4] = xformed

def imixColumns(block: bytearray):
    for i in range(4):
        col = block[4*i:4*i + 4]
        xformed = 4 * [0]

        xformed[0] = gmul(col[0], 14) ^ gmul(col[1], 11) ^ gmul(col[2], 13) ^ gmul(col[3],  9)
        xformed[1] = gmul(col[0],  9) ^ gmul(col[1], 14) ^ gmul(col[2], 11) ^ gmul(col[3], 13)
        xformed[2] = gmul(col[0], 13) ^ gmul(col[1],  9) ^ gmul(col[2], 14) ^ gmul(col[3], 11)
        xformed[3] = gmul(col[0], 11) ^ gmul(col[1], 13) ^ gmul(col[2],  9) ^ gmul(col[3], 14)

        block[4*i:4*i + 4] = xformed

### Core ###

def expand_key(key: bytes) -> bytes:
    num_rounds = 10
    rcons = [bytearray(4 * [0]) for _ in range(num_rounds)]

    rcons[0][0] = 1
    for i in range(1, num_rounds):
        rcons[i][0] = gmul(rcons[i-1][0], 2)
        rcons[i] = bytes(rcons[i])

    def S_word(word):
        return bytes([Sbox[b] for b in word])

    def rot(word):
        return word[1:] + word[:1]

    def xor(w0, w1):
        return bytes([x ^ y for x, y in zip(w0, w1)])

    key = [key[:4], key[4:8], key[8:12], key[12:]]
    rkeys = []
    for i in range(4 * (num_rounds + 1)):
        if i < 4:
            rkeys.append(key[i])
        elif i % 4 == 0:
            rc = rcons[i // 4 - 1]
            rkeys.append(xor(S_word(rot(rkeys[i-1])), xor(rkeys[i-4],  rc)))
        else:
            rkeys.append(xor(rkeys[i-1], rkeys[i-4]))

    return b''.join(rkeys)

def enc_block(block: bytearray, key: bytes):
    rkeys = expand_key(key)

    for i in range(16):
        block[i] ^= rkeys[i]

    for rnd in range(1, 11):
        subBytes(block)
        shiftRows(block)
        if rnd != 10: mixColumns(block)

        for i in range(16):
            block[i] ^= rkeys[16*rnd + i]

def dec_block(block: bytearray, key: bytes):
    rkeys = expand_key(key)

    for rnd in range(10, 0, -1):
        for i in range(16):
            block[i] ^= rkeys[16*rnd + i]

        if rnd != 10: imixColumns(block)
        ishiftRows(block)
        isubBytes(block)

    for i in range(16):
        block[i] ^= rkeys[i]

### Modes ###

def add_padding(b: bytes) -> bytes:
    len_padding = 16 - len(b) % 16
    return b + len_padding * len_padding.to_bytes(1, 'little')

def ecb_enc(plaintxt: bytes, key: bytes) -> bytes:
    pt = bytearray(add_padding(plaintxt))

    for i in range(len(pt) // 16):
        block = pt[16*i:16*(i+1)]
        enc_block(block, key)
        pt[16*i:16*(i+1)] = block

    return bytes(pt)

def ecb_dec(ciphertxt: bytes, key: bytes) -> bytes:
    if len(ciphertxt) % 16:
        raise ValueError('incorrect padding')

    ct = bytearray(ciphertxt)

    for i in range(len(ct) // 16):
        block = ct[16*i:16*(i+1)]
        dec_block(block, key)
        ct[16*i:16*(i+1)] = block

    return bytes(ct)
